dali/pipeline/operators/arithmetic/expression_impl_gpu.cuh

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_ARITHMETIC_EXPRESSION_IMPL_GPU_CUH_
+#define DALI_PIPELINE_OPERATORS_ARITHMETIC_EXPRESSION_IMPL_GPU_CUH_
+
+#include "dali/core/any.h"
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/pipeline/operators/arithmetic/expression_impl_factory.h"
+#include "dali/pipeline/operators/arithmetic/expression_tree.h"
+#include "dali/pipeline/operators/op_spec.h"
+#include "dali/pipeline/util/backend2workspace_map.h"
+#include "dali/pipeline/workspace/workspace.h"
+
+namespace dali {
+
+template <ArithmeticOp op, typename Result, typename Left, typename Right>
+__global__ void ExecuteBinTT(Result *result, const Left *l, const Right *r, int64_t extent) {
+  using meta = arithm_meta<op>;
+  int64_t i = blockIdx.x * blockDim.x + threadIdx.x;
+  if (i < extent) {
+    result[i] = meta::impl(l[i], r[i]);
+  }
+}
+
+template <ArithmeticOp op, typename Result, typename Left, typename Right>
+__global__ void ExecuteBinST(Result *result, Left l, const Right *r, int64_t extent) {
+  using meta = arithm_meta<op>;
+  int64_t i = blockIdx.x * blockDim.x + threadIdx.x;
+  if (i < extent) {
+    result[i] = meta::impl(l, r[i]);
+  }
+}
+
+template <ArithmeticOp op, typename Result, typename Left, typename Right>
+__global__ void ExecuteBinTS(Result *result, const Left *l, Right r, int64_t extent) {
+  using meta = arithm_meta<op>;
+  int64_t i = blockIdx.x * blockDim.x + threadIdx.x;
+  if (i < extent) {
+    result[i] = meta::impl(l[i], r);
+  }
+}
+
+int GetBlockNum(int extent, int thread_num) {
+  return (extent + thread_num - 1) / thread_num;
+}
+
+
+// Assumes that it will get a workspace for given Backend,
+// and the backend will store the inputs and outputs at given Backend.
+template <ArithmeticOp op, typename Result,
```

Comment:
```suggestion
template <ArithmeticOp op, typename ResultType,
```

Reply:
I don't think adding `...Type` everywhere improves anything - and is more likely to cause excessive line length and hard-to-read line breaks.

Reply:
Fair enough :)